palace/linalg/densematrix.cpp

```diff
@@ -0,0 +1,280 @@
+    const double tlambda1 = functor(lambda1) / n1;
+    const double tlambda2 = functor(lambda2) / n2;
+    const double tlambda3 = functor(lambda3) / n3;
+
+    const double at = (tlambda1 * l1mcmem1 * l1mcmem1 + tlambda2 * l2mcmem2 * l2mcmem2 +
+                       tlambda3 * l3mcmem3 * l3mcmem3) /
+                      f2;
+    const double bt = tlambda1 * m1 * m1 + tlambda2 * m2 * m2 + tlambda3 * m3 * m3;
+    const double ct = tlambda1 + tlambda2 + tlambda3;
+    const double dt =
+        (tlambda1 * m1 * l1mcmem1 + tlambda2 * m2 * l2mcmem2 + tlambda3 * m3 * l3mcmem3) /
+        f;
+    const double et = tlambda1 * m1 + tlambda2 * m2 + tlambda3 * m3;
+    const double ft = (tlambda1 * l1mcmem1 + tlambda2 * l2mcmem2 + tlambda3 * l3mcmem3) / f;
+    Mout(0, 0) = at;
+    Mout(0, 1) = dt;
+    Mout(0, 2) = ft;
+    Mout(1, 0) = dt;
+    Mout(1, 1) = bt;
+    Mout(1, 2) = et;
+    Mout(2, 0) = ft;
+    Mout(2, 1) = et;
+    Mout(2, 2) = ct;
+    return Mout;
+  }
+  else
+  {
+    MFEM_ABORT("MatrixFunction only supports 2x2 or 3x3 matrices!");
+  }
+  return Mout;
+}
+
+}  // namespace
+
+namespace linalg
+{
+
+mfem::DenseMatrix MatrixSqrt(const mfem::DenseMatrix &M)
+{
+  return MatrixFunction(M, [](auto s) { return std::sqrt(s); });
+}
+
+mfem::DenseTensor MatrixSqrt(const mfem::DenseTensor &T)
+{
+  mfem::DenseTensor S(T);
+  for (int k = 0; k < T.SizeK(); k++)
+  {
+    S(k) = MatrixSqrt(T(k));
+  }
+  return S;
+}
+
+mfem::DenseMatrix MatrixPow(const mfem::DenseMatrix &M, double p)
+{
+  return MatrixFunction(M, [p](auto s) { return std::pow(s, p); });
+}
+
+mfem::DenseTensor MatrixPow(const mfem::DenseTensor &T, double p)
+{
+  mfem::DenseTensor S(T);
+  for (int k = 0; k < T.SizeK(); k++)
+  {
+    S(k) = MatrixPow(T(k), p);
+  }
+  return S;
+}
+
+double SingularValueMax(const mfem::DenseMatrix &M)
+{
+  MFEM_ASSERT(
+      M.Height() == M.Width() && M.Height() > 0 && M.Height() <= 3,
+      "Matrix singular values only available for square matrices of dimension <= 3!");
+  const int N = M.Height();
+  if (N == 1)
+  {
+    return M(0, 0);
+  }
+  else if (N == 2)
+  {
+    return mfem::kernels::CalcSingularvalue<2>(M.Data(), 0);
+  }
+  else
+  {
+    return mfem::kernels::CalcSingularvalue<3>(M.Data(), 0);
+  }
+}
+
+double SingularValueMin(const mfem::DenseMatrix &M)
+{
+  MFEM_ASSERT(
+      M.Height() == M.Width() && M.Height() > 0 && M.Height() <= 3,
+      "Matrix singular values only available for square matrices of dimension <= 3!");
+  const int N = M.Height();
+  if (N == 1)
+  {
+    return M(0, 0);
+  }
+  else if (N == 2)
+  {
+    return mfem::kernels::CalcSingularvalue<2>(M.Data(), 1);
+  }
+  else
+  {
+    return mfem::kernels::CalcSingularvalue<3>(M.Data(), 2);
+  }
+}
+
+mfem::DenseTensor Mult(const mfem::DenseTensor &A, const mfem::DenseTensor &B)
+{
+  MFEM_VERIFY(A.SizeK() == B.SizeK(),
+              "Size mismatch for product of two DenseTensor objects!");
+  mfem::DenseTensor C(A);
```

Comment:
Doesn't sizing `C` like that assume that A and B both have the same size for the other two dimensions? I.e. that A and B are square aside from the K direction. Doesn't matter for the use I found, where they are square, but for general use it would.

Reply:
[dc0eb1b](https://github.com/awslabs/palace/pull/233/commits/dc0eb1b02c3f53e325c6be08c3c20b973e0275d6)